### PR TITLE
[SOAR-8945] Microsoft Log Analytics - Trigger

### PR DESCRIPTION
**Pull request by igorski-r7**
### Description

Describe the proposed changes:

  - Added Search Trigger 

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [X] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [X] Screenshot of job output with the plugin changes
- [X] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [X] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [X] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [X] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [X] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [X] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [X] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [X] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [X] Work fully completed
- [X] Functional
  - [X] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [X] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [X] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [X] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [X] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Test

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7_test/Microsoft Log Analytics:0.1.4. Step name: search\nUpdating auth token...\nAuthentication Token: ****************B6WRA\n",
    "status": "ok",
    "meta": {},
    "output": {
      "success": true
    }
  },
  "version": "v1",
  "type": "trigger_event"
}
```

<summary>
docker run --rm -i rapid7/microsoft_log_analytics:1.0.0 --debug test < tests/search.json
</summary>
</details>

### Run

<details>

```
{
  "body": {
    "meta": {},
    "output": {
      "tables": [
        {
          "name": "PrimaryResult",
          "columns": [
            {
              "name": "TenantId",
              "type": "string"
            },
            {
              "name": "SourceSystem",
              "type": "string"
            },
            {
              "name": "MG",
              "type": "string"
            },
            {
              "name": "ManagementGroupName",
              "type": "string"
            },
            {
              "name": "TimeGenerated",
              "type": "datetime"
            },
            {
              "name": "Computer",
              "type": "string"
            },
            {
              "name": "RawData",
              "type": "string"
            },
            {
              "name": "parameter_1_s",
              "type": "string"
            },
            {
              "name": "Type",
              "type": "string"
            },
            {
              "name": "_ResourceId",
              "type": "string"
            }
          ],
          "rows": [
            {
              "TenantId": "3a0309f6-5777-4792-bc91-3807512d2a2a",
              "SourceSystem": "RestAPI",
              "MG": "",
              "ManagementGroupName": "",
              "TimeGenerated": "2022-03-31T08:58:33.128Z",
              "Computer": "",
              "RawData": "",
              "parameter_1_s": "Test13",
              "Type": "Test1Log1_CL",
              "_ResourceId": ""
            },
            {
              "TenantId": "3a0309f6-5777-4792-bc91-3807512d2a2a",
              "SourceSystem": "RestAPI",
              "MG": "",
              "ManagementGroupName": "",
              "TimeGenerated": "2022-03-31T09:53:31.403Z",
              "Computer": "",
              "RawData": "",
              "parameter_1_s": "Test14",
              "Type": "Test1Log1_CL",
              "_ResourceId": ""
            }
          ]
        }
      ]
    },
    "log": "Connect: Connecting...\nrapid7_test/Microsoft Log Analytics:0.1.4. Step name: search\n\nRunning query Test1Log1_CL | where TimeGenerated > ago(7200s)...\nUpdating auth token...\nAuthentication Token: ****************Qrw6g\nGetting shared key...\nShared Key: ****************BHQ==\nGetting workspace ID...\nWorkspace ID: ****************d2a2a\nUpdating auth token...\nAuthentication Token: ****************w557w\nResponse received!\n"
  },
  "type": "trigger_event",
  "version": "v1"
}
```

<summary>
docker run --rm -i rapid7/microsoft_log_analytics:1.0.0 --debug run < tests/search.json
</summary>
</details>

### Icon-validate

<details>

```
[*] Validating plugin with all validators at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator SupportedVersionValidator
[*] Executing validator UnapprovedKeywordsValidator
[*] Executing validator HelpExampleValidator
[*] Executing validator Version Increment Validator
[*] Executing validator ExceptionValidator
[*] Executing validator CredentialsValidator
[*] Executing validator PasswordValidator
[*] Executing validator PrintValidator
[*] Executing validator ConfidentialValidator
[*] Executing validator DockerValidator
[*] Executing validator URLValidator
[*] Plugin failed validation! The following validation errors occurred:

Validator "HelpExampleValidator" failed! 
        Cause: Invalid JSON example identified in Example input of action titled Input. Please correct the JSON example in help.md.
        Improperly formatted JSON example identified in Example input of action titled Input. Please updated the JSON example in help.md with 2 space indentation.


----
[*] Total time elapsed: 17005.576999999997ms

```

<summary>
icon-validate --all .
</summary>
</details>

### UI
Search Trigger job output

SearchTrigger1

### CODE REVIEW

File: plugins/microsoft_log_analytics/plugin.spec.yaml [hunk 1]
```diff
@@ -141,6 +141,48 @@ actions:
         type: string
         required: true
         example: "AzureActivity | summarize count() by Category"
+    output:
+      tables:
+        title: Tables
+        description: Array of tables representing the query result. Each table contains name, columns, and rows keys
```
Comment (cmcnally-r7) [inline plugins/microsoft_log_analytics/plugin.spec.yaml:147]: https://docs.rapid7.com/insightconnect/style-guide#punctuation
Try changing the description to:
`Array of tables representing the query result, with each table containing a name, column and row keys`

**Review by cmcnally-r7 — CHANGES_REQUESTED** [review]

**Review by cmcnally-r7 — APPROVED** [review]

File: plugins/microsoft_log_analytics/plugin.spec.yaml [hunk 2]
```diff
@@ -144,6 +144,48 @@ actions:
...
+        description: Customer's workspace name that the application registration portal is assigned
+        type: string
+        required: true
+        example: ExampleWorkspaceName
+      query:
+        title: Query
+        description: Microsoft Log Analytics query
+        type: string
+        required: true
+        example: "AzureActivity | summarize count() by Category"
```
Comment (ablakley-r7) [inline plugins/microsoft_log_analytics/plugin.spec.yaml:185]: The '|' in the example will cause this to format incorrectly in the help.md. Is there an example query that doesn't have the '|' character?

File: plugins/microsoft_log_analytics/icon_microsoft_log_analytics/triggers/search/trigger.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+            response = return_non_empty_query_output(
+                self.connection.client.get_log_data(subscription_id, resource_group_name, workspace_name, query)
+            )
+            if response.get("tables"):
+                self.logger.info("Response received!")
+                self.send(response)
+            else:
+                self.logger.info("No response has been received!")
+            self.logger.info(f"Sleeping for {interval} seconds...")
+            time.sleep(interval)
```
Comment (ablakley-r7) [inline plugins/microsoft_log_analytics/icon_microsoft_log_analytics/triggers/search/trigger.py:32]: May be overthinking this but should there be a requirement to handle negative integer input for interval? To either raise a plugin exception and exit gracefully or default to another value. Might be worth a discussion in how we handle these inputs overall.

**Review by ablakley-r7 — CHANGES_REQUESTED** [review]

File: plugins/microsoft_log_analytics/icon_microsoft_log_analytics/triggers/search/trigger.py (hunk 3 above)
Comment (igorski-r7) [reply plugins/microsoft_log_analytics/icon_microsoft_log_analytics/triggers/search/trigger.py:32]: I was thinking about that and I see two options:
1. Just an absolute value from that integer.
2. Raise PluginException to handle negative numbers

I think the first one would be much better than second 😄

File: plugins/microsoft_log_analytics/plugin.spec.yaml (hunk 2 above)
Comment (igorski-r7) [reply plugins/microsoft_log_analytics/plugin.spec.yaml:185]: Hard to get some better queries without "|" character. I think I'll stay with "AzureActivity" alone.

File: plugins/microsoft_log_analytics/plugin.spec.yaml [hunk 4]
```diff
@@ -140,10 +140,52 @@ actions:
...
+    title: Search
+    description: Run Log Analytics query every interval time (expressed in seconds)
+    input:
+      interval:
+        title: Interval
+        description: Integer value that represents interval time in seconds
+        type: integer
+        required: true
+        example: 10
+        default: 10
```
Comment (blaxminarayan-r7) [inline plugins/microsoft_log_analytics/plugin.spec.yaml:161]:
```suggestion
        default: 60
```

File: plugins/microsoft_log_analytics/plugin.spec.yaml [hunk 5]
```diff
@@ -140,10 +140,52 @@ actions:
...
+        description: Customer's workspace name that the application registration portal is assigned
+        type: string
+        required: true
+        example: ExampleWorkspaceName
+      query:
+        title: Query
+        description: Microsoft Log Analytics query
+        type: string
+        required: true
+        example: "AzureActivity I summarize count() by Category"
```
Comment (blaxminarayan-r7) [inline plugins/microsoft_log_analytics/plugin.spec.yaml:185]:
```suggestion
        example: "AzureActivity | summarize count() by Category"
```

**Review by blaxminarayan-r7 — CHANGES_REQUESTED** [review]

**Review by blaxminarayan-r7 — APPROVED** [review]

**Pull request merged**